src/index/bip352.cpp

```diff
@@ -0,0 +1,150 @@
+{
+public:
+    explicit DB(fs::path file_name, size_t n_cache_size, bool f_memory = false, bool f_wipe = false);
+
+    bool WriteSilentPayments(const std::pair<uint256, tweak_index_entry>& kv);
+};
+
+BIP352Index::DB::DB(fs::path file_name, size_t n_cache_size, bool f_memory, bool f_wipe) :
+    BaseIndex::DB(gArgs.GetDataDirNet() / "indexes" / file_name, n_cache_size, f_memory, f_wipe)
+{}
+
+bool BIP352Index::DB::WriteSilentPayments(const std::pair<uint256, tweak_index_entry>& kv)
+{
+    CDBBatch batch(*this);
+    batch.Write(std::make_pair(DB_SILENT_PAYMENT_INDEX, kv.first), kv.second);
+    return WriteBatch(batch);
+}
+
+BIP352Index::BIP352Index(bool cut_through, std::unique_ptr<interfaces::Chain> chain, size_t n_cache_size, bool f_memory, bool f_wipe)
+    : BaseIndex(std::move(chain), strprintf("bip352 %sindex", cut_through ? "cut-through " : ""), /*start_height=*/Params().IsTestChain() ? 0 : TAPROOT_MAINNET_ACTIVATION_HEIGHT), m_db(std::make_unique<BIP352Index::DB>(fs::u8path(strprintf("bip352%s", cut_through ? "ct" : "")), n_cache_size, f_memory, f_wipe))
+{
+    m_cut_through = cut_through;
+}
+
+BIP352Index::~BIP352Index() = default;
+
+bool BIP352Index::GetSilentPaymentKeys(std::vector<CTransactionRef> txs, CBlockUndo& block_undo, tweak_index_entry& index_entry)
+{
+    assert(txs.size() - 1 == block_undo.vtxundo.size());
+
+    for (uint32_t i=0; i < txs.size(); i++) {
+        auto& tx = txs.at(i);
+
+        if (!BIP352::MaybeSilentPayment(tx)) continue;
+
+        // -1 as blockundo does not have coinbase tx
+        CTxUndo undoTX{block_undo.vtxundo.at(i - 1)};
+        std::map<COutPoint, Coin> coins;
+
+        for (uint32_t j = 0; j < tx->vin.size(); j++) {
+            coins[tx->vin.at(j).prevout] = undoTX.vprevout.at(j);
+        }
+
+        std::optional<CPubKey> tweaked_pk = BIP352::GetSerializedSilentPaymentsPublicData(tx->vin, coins);
+        if (tweaked_pk) {
+            // Used to filter dust. To keep the index small we use only one byte
+            // and measure in hexasats.
+            uint8_t max_output_hsat = 0;
+            for (const CTxOut& txout : tx->vout) {
+                if (!txout.scriptPubKey.IsPayToTaproot()) continue;
+                uint8_t output_hsat = txout.nValue > max_dust_threshold ? UINT8_MAX : txout.nValue >> dust_shift;
+                max_output_hsat = std::max(output_hsat, max_output_hsat);
+            }
+
+            if (m_cut_through) {
+                // Skip entry if all outputs have been spent.
+                // This is only effective when the index is generated while
+                // the tip is far ahead.
+                //
+                // This is done after calculating the tweak in order to minimize
+                // the number of UTXO lookups.
+                LOCK(cs_main);
+                const CCoinsViewCache& coins_cache = m_chainstate->CoinsTip();
+
+                uint32_t spent{0};
+                for (uint32_t i{0}; i < tx->vout.size(); i++) {
```

Comment:
This loop index shadows an outer variable `i`. Also should be using `size_t` here.